webui/react/src/e2e/tests/projectsWorkspaces.spec.ts

```diff
@@ -0,0 +1,125 @@
+import { expect } from '@playwright/test';
+import { v4 } from 'uuid';
+
+import { test } from 'e2e/fixtures/global-fixtures';
+import { WorkspaceCreateModal } from 'e2e/models/components/WorkspaceCreateModal';
+import { Workspaces } from 'e2e/models/pages/Workspaces';
+
+test.describe('Projects', () => {
+  let wsCreatedWithButton: string = '';
+  let wsCreatedWithSidebar: string = '';
+  const createWorkspaceAllFields = async function (
+    modal: WorkspaceCreateModal,
+    wsNamePrefix: string,
+  ): Promise<string> {
+    const uuid = v4();
+    const fullName = `${wsNamePrefix}-${uuid}`;
+    await modal.workspaceName.pwLocator.fill(fullName);
+
+    const randId = () => Math.floor(Math.random() * 10).toString();
+    await modal.useAgentUser.switch.pwLocator.click();
+    await modal.agentUid.pwLocator.fill(randId());
+    await modal.agentUser.pwLocator.fill(`user-${uuid}`);
+
+    await modal.useAgentGroup.switch.pwLocator.click();
+    await modal.agentGid.pwLocator.fill(randId());
+    await modal.agentGroup.pwLocator.fill(`group-${uuid}`);
+
+    await modal.useCheckpointStorage.switch.pwLocator.click();
+    await expect(modal.checkpointCodeEditor.pwLocator).toBeVisible();
+    await modal.useCheckpointStorage.switch.pwLocator.click();
+    await expect(modal.checkpointCodeEditor.pwLocator).not.toBeVisible();
+
+    await modal.footer.submit.pwLocator.click();
+    await expect(modal.pwLocator).not.toBeVisible();
+    return fullName;
+  };
+
+  test.beforeEach(async ({ dev, auth, page }) => {
+    await dev.setServerAddress();
+    await auth.login(/dashboard/);
+    await expect(page).toHaveTitle('Home - Determined');
+    await expect(page).toHaveURL(/dashboard/);
+  });
+
+  test.afterEach(async ({ page }) => {
+    const workspacesPage = new Workspaces(page);
+    await test.step('Delete a workspace', async () => {
```

Comment:
best not to have split behavior in after all.
i would say the afterAll should delete any workspace through a single method. if we want to test an alternate method for deleting workspaces, then we should perform that in a test case. Check out how I deactivate users at the end of user tests.

Reply:
> best not to have split behavior in after all.
Why?
I need to test both ways of deleting. It's explicit in the test plan. I actually kind of like having steps in after all as clean-up, but I'm planning to defer the big test-organizational stuff until a future PR. 

There's a trade off here, we can split everything into separate test cases which is cleaner, but also slower and more duplicative since we need to create/delete more workspaces overall. This test case needs 2 workspaces, so doing create/delete differently with them both to get the free coverage is better IMO than adding an additional test case to get the same coverage. I suspect at the end I may need to split this into separate cases. It's a big test. But for now I'm planning to leave it to see if we can have that coverage without more test cases.

Reply:
intuitively, I would think if teardown should be the most stable and straightforward part of the test. i.e., if it fails on one test, i would expect it to fail in every test. if i'm just running a single test, it could be confusing which teardown is running. if we want two teardowns, maybe we should have two describes. anyway, this is all semantics and ultimately these details dont matter so much. I'll leave the decision to you